### DISCUSSION TITLE
[Idea] Support Factur-X in pdf invoices

### DISCUSSION THREAD

**Discussion by jf-viguier** [PrestaShop/PrestaShop#30844] [Ideas]
e-invoice is soon mandatory by law in Europe.

**Comment by kpodemski** [comment]
Hi @jf-viguier 

I think you provided too few details to understand the need. Many merchants use external services to handle invoicing. How would you like to help merchants support e-invoice in PrestaShop?

**Reply by jf-viguier** [reply]
**Obligations**
At european level :
https://ec.europa.eu/digital-building-blocks/wikis/display/DIGITAL/European+legislation+on+eInvoicing
in France :
https://www.economie.gouv.fr/cedef/facturation-electronique-entreprises

**About Factur-X**
Information about the Factur-X http://fnfe-mpe.org/factur-x/
It's a format with PDF/A-3 https://en.wikipedia.org/wiki/PDF/A#PDF/A-3
A php lib here : https://github.com/atgp/factur-x

**Comment by Patrickclicweb** [comment]
Hello,
If there are users of Prestashop who have a CRM or ERP is for my agency not the case of most of them 

I think that the management of PDFs with the obligation to invoice-X can not be done because often the rounding rules on products with low value and sold in large quantities the different VAT is not compliant in BtoB I have the case of products with declination that have 1 or ,cent more or less than the initial price and on top I have groups of customers who have a discount of 5, 7 or 9% or more and it is often products purchased by 25, 50 or 100 pieces at once 
Moreover with the Cart Rules this is even more true 
I think it is not serious to tell a "small" customer to take an ERP, how the Prestashop edition support will manage this? 
Thank

**Reply by kpodemski** [reply]
Price rounding/Cart Rules are currently being analyzed for future PrestaShop versions. It's a complex topic where the project could use help from the community.

>I think it is not serious to tell a "small" customer to take an ERP, how the Prestashop edition support will manage this? 

I think that's the question for the Edition team 👍🏻

**Reply by Patrickclicweb** [reply]
For Prestashop edition it was an example I have 25 customers including 3 still under 1.6.1.24 and just 6 who have ERP 
Is it possible to mention here the Prestashop Edition team? 
The others, how to explain to them that Prestashop an Ecommerce software is not compliant and that they have to take another tool?

**Reply by Patrickclicweb** [reply]
For the cart Rules I participated to 2 events with Camille with FOP and I'm afraid that it ends with a simple refacto :(

**Reply by kpodemski** [reply]
> For the cart Rules I participated to 2 events with Camille with FOP

That's awesome! Thanks for contributing! The more people discuss here and vote, the more visibility subject will have.

**Comment by Progi1984** [comment]
A library exists : https://github.com/atgp/factur-x

**Comment by Hlavtox** [comment]
I think we should not focus on region-specific restrictions and laws, but provide a perfect support to implement them from modules.

Currently, the only way to change generated PDFs is by an override.

We should be able to EASILY implement a module to provide own custom way of rendering invoices, and allow to fetch them externally from ERPs.

**Reply by Patrickclicweb** [reply]
Hi,
And I think the opposite! 
We have to focus on these things to avoid small users to add an ERP 
I don't see how to explain it to them? 
I migrated a lot of customer (small merchant) from Woocommerce to Prestashop, but with this new done, I think I'll do the reverse if this request does not succeed 
It is unthinkable for me that Prestashop can't do without an ERP to be in conformity

**Reply by Hlavtox** [reply]
It would be the ideal world, but this is impossible. Every country, continent, has their own specific rules and laws. We would need to have million different possibilities. :(

**Reply by MatShir** [reply]
I don't see why @Hlavtox and @Patrickclicweb views are incompatible. We can make the software more modular/flexible and in the sometime, we could provide basic need to make it (in the core or native module). 
_Maybe, it could be on the topic of the next product council to be discussed._

**Reply by Patrickclicweb** [reply]
@MatShir 
Why does PrestaShop not offer a native ERP like with Prestashop Checout PrestaShop Assurance Prestashop marketing
not enough monetization possible?

**Reply by MatShir** [reply]
It seems a good idea. I don't know why! Maybe they didn't find the right partnership or they prefer to rely on addons modules connectors, who knows

**Reply by Patrickclicweb** [reply]
I wonder how. does "PrestaShop edition" work in this context? :(

**Comment by Prestaplugins** [comment]
For having already worked with factur-X and its PHP library, the most important thing is really to have no approximation in the calculation of the amounts.
It is necessary for example that the total HT / TTC of the invoice or the total of the VAT is absolutely equal to the sum of the HT / TTC / VAT of each line of the invoice, otherwise the XML document is invalid.
So no cents more or less here and there...

**Reply by jeckyl** [reply]
This confirms that no PrestaShop invoice with a discount will be valid at this time.
![image](https://github.com/PrestaShop/PrestaShop/assets/595633/035d8406-46bf-499a-bd0b-e75c52b9760c)
![image (2)](https://github.com/PrestaShop/PrestaShop/assets/595633/9170622a-da2b-4b80-aa1c-5990f967afd4)

**Reply by Patrickclicweb** [reply]
Yes all the calculation methods in this case are wrong

**Comment by marsaldev** [comment]
In Italy is mandatory since 3 years (more or less), with 2 extra fields necessary for invoicing, SDI and/or PEC e-mail.
"We" solved with a module (different), most of merchants abandoned the PrestaShop invoicing and linked an external ERP or invoicing software. 
As @Hlavtox said, each country is different, it's impossible to include in the core this feature, it's more easy develop a specific module and manage the invoicing externally.

**Reply by MatShir** [reply]
I didn't look in detail, but if some are required for a lot of countries we could make a native module.

**Reply by jf-viguier** [reply]
Yes, the entire Europe is concerned

**Reply by luigimassa** [reply]
In addons for Italy electronic invoice there is this module https://addons.prestashop.com/en/price-management/49979-vat-number-and-personal-tax-code-pro-pec-sia.html 
This module add 2 mandatory fields: PEC and SDI

I think it is necessary to list all the requirements of all individual countries and see which fields are mandatory
Electronic invoicing is merchant specific and each merchant will want to use their own channel. By now we know a lot about it here in Italy
Another problem to consider is sending the invoice between countries of the European community, but I know that the EU is still creating the specifications

**Reply by jf-viguier** [reply]
Thanks @luigimassa. What is PEC and SIA ?

**Reply by marsaldev** [reply]
@jf-viguier The PEC is a certified e-mail that have legal value, I think Luigi wants to wrote SDI, it's a unique identifier for companies that provide invoicing/documents archiving and transmitting data to the government on your behalf.

**Comment by yannicka** [comment]
Where are we on this topic?

**Reply by Hlavtox** [reply]
We are currently at “prestashop calculates total products 2007 CZK + shipping 0 CZK = total order 2008 CZK”. (https://github.com/PrestaShop/PrestaShop/issues/36369) And Prestashop company does everything else other than fix these basic things.

So if you want an ecommerce system supporting something like this topic before you are in a coffin, you must implement it yourself. 😁

**Reply by jf-viguier** [reply]
You are harsh @Hlavtox !

**Reply by Hlavtox** [reply]
@jf-viguier That's life!

**Reply by jolelievre** [reply]
Just to remind you, the pricing computing topic is planned for the coming version
We can't say for sure when it will be finished (because the topic is HUGE) and which version will integrate it as a stable feature

But we are gonna start a big refacto on this once we're done with Discounts What I hope is that we will implement it as an experimental feature in the coming 9.x versions, and hopefully it will be over for the 10.0

**Reply by jf-viguier** [reply]
Thanks @jolelievre !

**Reply by Patrickclicweb** [reply]
@jolelievre Aurons nous de la rétro-compatibilité sur ce sujet pour la V8 ?
C'est un sujet bloquant pour beaucoup de nos clients

**Reply by kpodemski** [reply]
@Patrickclicweb 

PrestaShop 8 won't include any built-in options for this requirement. However, you can likely find suitable solutions within the ecosystem. 

PS: Please keep conversation in English language in this space.

**Comment by KaeligBrl** [comment]
Hello,

I’m currently running a PrestaShop store for a business based in France, and with the upcoming mandatory e-invoicing reform (2026–2027), I’d like to understand how this will be handled within the CMS.

Could you please let me know:

Whether PrestaShop has already implemented (or plans to implement) native support for French e-invoicing requirements

If there is a recommended module or official integration with certified PDP / e-invoicing platforms

The current roadmap or any updates regarding compliance with the French regulation

I’d like to anticipate the necessary technical setup to remain compliant, so any guidance or documentation would be greatly appreciated.

Thank you in advance for your help.

Best regards,

Kaëlig

**Reply by Eolia** [reply]
See here https://bb.enter-solutions.net/topic/1269/facturation-%C3%A9lectronique-2026-existe-t-il-un-vrai-module-phenix-pa-automatique/4

**Reply by sventorfs** [reply]
@KaeligBrl Hi, I have recently built a simple PrestaShop module for a client for e-invoicing. It does not have a ton of features but just allows you to export the invoices in the mandatory EN 16931 format and send them to your accountant or upload them in your accounting software. (We are working on direct PA-integration but the code is not ready yet.)
 
Here is the link if you would be interested to take a look: https://addons.prestashop.com/en/product-management/96878-peppol-ubl-downloader.html
